driver/src/metrics/http_metrics.rs

```diff
@@ -0,0 +1,249 @@
+use anyhow::Result;
+use prometheus::{IntGauge, IntGaugeVec, Opts, Registry};
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+use std::fmt::{self, Debug, Formatter};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// A registry for all HTTP related metrics.
+#[derive(Debug)]
+pub struct HttpMetrics {
+    labeled: HashMap<TypeId, SubsystemMetrics<IntGaugeVec>>,
+    unlabeled: HashMap<TypeId, SubsystemMetrics<IntGauge>>,
+}
+
+/// Metrics related to a single HTTP subsystem. Subsystems are grouped based on
+/// what the HTTP client is used for.
+pub struct SubsystemMetrics<Gauge> {
+    latency: Gauge,
+    size: Gauge,
+}
+
+impl<Gauge> Debug for SubsystemMetrics<Gauge> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_struct("SubsystemMetrics").finish()
+    }
+}
+
+impl HttpMetrics {
+    /// Create a new HTTP metrics registry.
+    pub fn new(registry: &Arc<Registry>) -> Result<Self> {
+        let mut metrics = HttpMetrics {
+            labeled: HashMap::new(),
+            unlabeled: HashMap::new(),
+        };
+
+        metrics.initialize_labeled_subsystem::<Web3Label>(registry)?;
+        metrics.initialize_labeled_subsystem::<KrakenLabel>(registry)?;
+        metrics.initialize_unlabeled_subsystem::<GasStationLabel>(registry)?;
+
+        Ok(metrics)
+    }
+
+    /// Add a request latency and size measurement to the current HTTP metrics
+    /// registry for the specified label.
+    pub fn request_labeled<L: LabeledSubsystem + Any>(
+        &self,
+        label: L,
+        latency: Duration,
+        size: usize,
+    ) {
+        let subsystem = self
+            .labeled
+            .get(&TypeId::of::<L>())
+            .expect("all labeled subsystems have registered metrics");
+
+        subsystem
+            .latency
+            .with_label_values(&[label.label()])
+            .set(latency.as_millis() as _);
+
+        subsystem
+            .size
+            .with_label_values(&[label.label()])
+            .set(size as _);
+    }
+
+    /// Add a request latency and size measurement to the current HTTP metrics
+    /// registry.
+    pub fn request_unlabeled<L: UnlabeledSubsystem + Any>(&self, latency: Duration, size: usize) {
```

Comment:
Is it worth it to have both labeled and unlabeled? Unlabeled can be implemented in terms of a labeled with only one request type. Has the disadvantage that code using this has to pass `MySubsystem::default_or_something_like_that` but makes the code here simpler and then we don't have duplicate `_unlabeled` methods.

Reply:
I thought of that, but the `labeled` variants require a value which doesn't make sense for unlabeled subsystems.